gobblin-config-management/gobblin-config-client/src/main/java/gobblin/config/client/ConfigClient.java

```diff
@@ -0,0 +1,352 @@
+        return null;
+      }
+      
+      // both are absolute URI
+      if(floorKey.getAuthority().equals(configKeyURI.getAuthority())){
+        if(configKeyURI.getPath().startsWith(floorKey.getPath())) {
+          return floorKey;
+        }
+      }
+    }
+    
+    return null;
+  }
+  
+  private ConfigStoreAccessor createNewConfigStoreAccessor(URI configKeyURI) throws ConfigStoreFactoryDoesNotExistsException,
+  ConfigStoreCreationException, VersionDoesNotExistException{
+    
+    ConfigStoreAccessor result;
+    ConfigStoreFactory<ConfigStore> csFactory = this.getConfigStoreFactory(configKeyURI);
+    ConfigStore cs = csFactory.createConfigStore(configKeyURI);
+
+    if (!(cs instanceof ConfigStoreWithStableVersioning)) {
+      if (this.policy == VersionStabilityPolicy.CROSS_JVM_STABILITY) {
+        throw new RuntimeException(String.format(
+            "with policy set to %s, connect not connect to unstable config store %s",
+            VersionStabilityPolicy.CROSS_JVM_STABILITY, cs.getStoreURI()));
+      }
+    }
+    
+    String currentVersion = cs.getCurrentVersion();
+    // topology related
+    ConfigStoreBackedTopology csTopology = new ConfigStoreBackedTopology(cs, currentVersion);
+    InMemoryTopology inMemoryTopology = new InMemoryTopology(csTopology);
+    
+    // value related
+    ConfigStoreBackedValueInspector rawValueInspector = new ConfigStoreBackedValueInspector(cs, currentVersion, inMemoryTopology);
+    InMemoryValueInspector inMemoryValueInspector;
+    
+    // ConfigStoreWithStableVersioning always create Soft reference cache
+    if ( cs instanceof ConfigStoreWithStableVersioning){
+      inMemoryValueInspector = new InMemoryValueInspector(rawValueInspector, false);
+      result = new ConfigStoreAccessor(cs, inMemoryValueInspector, inMemoryTopology);
+    }
+    // Non ConfigStoreWithStableVersioning but require STRONG_LOCAL_STABILITY, use Strong reference cache
+    else if (this.policy == VersionStabilityPolicy.STRONG_LOCAL_STABILITY) {
+      inMemoryValueInspector = new InMemoryValueInspector(rawValueInspector, true);
+      result = new ConfigStoreAccessor(cs, inMemoryValueInspector, inMemoryTopology);
+    }
+    // Require No cache
+    else {
+      result = new ConfigStoreAccessor(cs, rawValueInspector, inMemoryTopology);
+    }
+    
+    return result;
+  }
+  
+  private ConfigStoreAccessor getConfigStoreAccessor(URI configKeyURI) throws ConfigStoreFactoryDoesNotExistsException,
+      ConfigStoreCreationException, VersionDoesNotExistException {
+    
+    URI matchedFloorKey = getMatchedFloorKeyFromCache(configKeyURI);
+    ConfigStoreAccessor result;
+    if(matchedFloorKey!=null){
+      result = this.configStoreAccessorMap.get(matchedFloorKey);
+      return result;
+    }
+    
+    result = createNewConfigStoreAccessor(configKeyURI);
+    ConfigStore cs = result.configStore;
+    
+    // put to cache
+    this.configStoreAccessorMap.put(cs.getStoreURI(), result);
```

Comment:
What happens if `cs.getStoreURI()` is not an ancestor of `configKeyURI`? Then repeatedly calling `getConfig(configKeyURI)` would instantiate a new store every time.
Our contract should probably by that `createNewConfigStoreAccessor(configKeyURI).configStore.getStoreURI()` must be an ancestor of `configKeyURI`, and explicitly check for this condition on the code.